ListView of activities - slidable actions

punker76 opened the pull request:
This PR adds two actions to the activity list.

- Add the UWP Community Toolkit, so we can use the SlidableListItem for this stuff
- Left command selects activity and make the Kudos and/or open the kudos list
- Right command selects activity and opens the comment input dialog (and after this the comment list)
- Also fix not showing comment pivot after swipe it
- Block giving Kudos on own activities

Closes #23 ListView of activities - use swipe to 'like'

punker76 committed:
Add Microsoft.Toolkit.Uwp.UI.Controls 1.1.0
  So now we can use the SlidableListItem

punker76 committed:
Use SlidableListItem
  Show also visual status for Kudos or not and the Comment icon

punker76 committed:
Add functionality to slidable items
  - Left command selects activity and make the Kudos and/or open the kudos list
  - Right command selects activity and opens the comment input dialog (and after this the comment list)

Depechie commented on src/Kliva/ViewModels/ActivityDetailViewModel.cs:
```diff
@@ -194,10 +197,10 @@ public ActivityDetailViewModel(INavigationService navigationService, IStravaServ
+                await _stravaService.PostComment(activityId.ToString(), dialog.Description);
+                await LoadActivityDetails(activityId.ToString());
             }
+            ServiceLocator.Current.GetInstance<IMessenger>().Send<PivotMessage<ActivityPivots>>(new PivotMessage<ActivityPivots>(ActivityPivots.Comments, true, true), Tokens.ActivityPivotMessage);
```
Shouldn't we only show the comments tab, if a comment was entered? So I would move this Send<> back into the if(){...}

punker76 replied:
Ok

Depechie commented on src/Kliva/ViewModels/MainViewModel.cs:
```diff
@@ -95,6 +95,34 @@ public ActivitySummary SelectedActivity
+            {
+                this.SelectedActivity = activitySummary;
+            }
+            MessengerInstance.Send<ActivitySummaryCommentMessage>(new ActivitySummaryCommentMessage(_selectedActivity));
```
Why use the internal backing field? Normally we would always tend to use the property... so MessengerInstance.Send<ActivitySummaryCommentMessage>(new ActivitySummaryCommentMessage(SelectedActivity));

punker76 replied:
My fault (copy&paste) will change it

Depechie commented on src/Kliva/XAMLResources/DataTemplates.xaml:
```diff
@@ -16,58 +17,77 @@
+                                             RightIcon="Comment"
+                                             RightLabel="Comment">
+            <interactivity:Interaction.Behaviors>
+                <core:DataTriggerBehavior Binding="{x:Bind HasKudoed, Mode=OneWay}" ComparisonCondition="Equal" Value="True">
```
This does not trigger 'after' given a Kudos. This only works when loading the list with items. Even setting it to TwoWay has no effect.
We need to see how we can fix this...

punker76 replied:
Yeah, cause ActivitySummary isn't updated after changes (it concerns also the other properties). For the first I can trigger the change of the property by hand.

Depechie replied:
Indeed I would suggest to manually search for the changed ActivitySummary inside the ActivityIncrementalCollection and update the property by hand if possible.

Depechie reviewed (commented):
Looking great so far, only the trigger after having given Kudos is a blocking issue right now.

bartlannoeye commented on src/Kliva/ViewModels/ActivityDetailViewModel.cs:
```diff
@@ -92,10 +92,10 @@ public bool IsEditEnabled
 
         private RelayCommand _kudosCommand;
-        public RelayCommand KudosCommand => _kudosCommand ?? (_kudosCommand = new RelayCommand(async () => await OnKudosAsync()));
+        public RelayCommand KudosCommand => _kudosCommand ?? (_kudosCommand = new RelayCommand(async () => await OnKudosAsync(this.SelectedActivity.Id)));
```
no need for using this.

Depechie replied:
Meaning the 'this' keyword

bartlannoeye commented on src/Kliva/ViewModels/MainViewModel.cs:
```diff
@@ -95,6 +95,34 @@ public ActivitySummary SelectedActivity
+        private RelayCommand<ActivitySummary> _commentCommand;
+        public RelayCommand<ActivitySummary> CommentCommand => _commentCommand ?? (_commentCommand = new RelayCommand<ActivitySummary>(async (x) => await OnCommentAsync(x)));
+
+        private async Task OnKudosAsync(ActivitySummary activitySummary)
```
private methods below public methods

bartlannoeye replied:
why async Task? I don't see async code

bartlannoeye commented on src/Kliva/ViewModels/MainViewModel.cs:
```diff
@@ -95,6 +95,34 @@ public ActivitySummary SelectedActivity
+        {
+            // make sure we work with the swiped item
+            // this could happen if we use the mouse to swipe and move the mouse outside the ListView and release it there
+            if (activitySummary != this.SelectedActivity)
```
no need for this.

punker76 replied:
> this could happen if we use the mouse to swipe and move the mouse outside the ListView and release it there

Depechie replied:
Bart means, no need for the 'this' keyword ;)
The actual code IS needed of course

punker76 replied:
uups

bartlannoeye replied:
meant the code this.SelectedActivity should simply be SelectedActivity, sorry for inclarity :)
I'm quite ocd on unused/unnecessary code ;)

bartlannoeye commented on src/Kliva/ViewModels/MainViewModel.cs:
```diff
@@ -95,6 +95,34 @@ public ActivitySummary SelectedActivity
+            MessengerInstance.Send<ActivitySummaryKudosMessage>(new ActivitySummaryKudosMessage(_selectedActivity));
+        }
+
+        private async Task OnCommentAsync(ActivitySummary activitySummary)
```
idem on async

punker76 committed:
Show the comments tab only if a comment was entered

punker76 committed:
Review changes
  Don't use backing field
  Remove `this`
  Remove unused Task result
  Move private methods below public methods

punker76 committed:
Review changes 2nd
  Remove another `this`
  Avoid null pointer exception on commands

Depechie commented on src/Kliva/ViewModels/MainViewModel.cs:
```diff
@@ -96,32 +96,10 @@ public ActivitySummary SelectedActivity
-            }
-            MessengerInstance.Send<ActivitySummaryCommentMessage>(new ActivitySummaryCommentMessage(_selectedActivity));
-        }
+        public RelayCommand<ActivitySummary> CommentCommand => _commentCommand ?? (_commentCommand = new RelayCommand<ActivitySummary>(OnCommentAsync));
```
Jan, Bart meant that you created 2 methods called OnKudosAsync and OnCommentAsync, but the code in these 2 methods don't contain any Async references... You only call MessengerInstance.Send<...>

So his suggestion : Rename the methods to OnKudos and OnComment!

punker76 committed:
Rename non async methods
  Sometimes I'm stupid...

punker76 committed:
This CanExecute doesn't work correctly

punker76 committed:
Give Kudos only if it's possible
